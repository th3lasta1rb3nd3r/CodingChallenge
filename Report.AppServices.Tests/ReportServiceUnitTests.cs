using AppServices;
using AppServices.Interfaces;
using FizzWare.NBuilder;
using Newtonsoft.Json;
using Reporting.Models;
using Snapshooter.Xunit;
using System;

namespace Report.AppServices.Tests
{
    public class ReportServiceUnitTests
    {
        private readonly IReportService _reportService;

        public ReportServiceUnitTests()
        {
            _reportService = new ReportService();
        }

        [Fact]
        public void GenerateReportInvalidXmlTest()
        {
            const string template = "<template-row>< firstname/></template-row>";

            var invalidTemplates = _reportService.ValidateTemplate<ReportModel>(template);

            Snapshot.Match(invalidTemplates);
        }

        [Fact]
        public void GenerateReportInvalidTemplateFieldTest()
        {
            const string template = "<template-row><firstname/><lastname1 /></template-row>";

            var invalidTemplates = _reportService.ValidateTemplate<ReportModel>(template);

            Snapshot.Match(invalidTemplates);
        }

        [Fact]
        public void GenerateReportSuccessTest()
        {
            var reports = Builder<ReportModel>.CreateListOfSize(5).Build();

            string json = JsonConvert.SerializeObject(reports);
            const string template = "<template-row><firstname/><lastname /></template-row>";

            var invalidTemplates = _reportService.ValidateTemplate<ReportModel>(template);

            if (invalidTemplates.Count > 0)
            {
                return;
            }

            var result = _reportService.GenerateReport(json, template);

            Snapshot.Match(result);
        }

        [Fact]
        public void GenerateReportSuccessDuplicateFieldsTest()
        {
            var reports = Builder<ReportModel>.CreateListOfSize(5).Build();

            string json = JsonConvert.SerializeObject(reports);
            const string template = "<template-row><firstname/><firstname /></template-row>";

            var invalidTemplates = _reportService.ValidateTemplate<ReportModel>(template);

            if (invalidTemplates.Count > 0)
            {
                return;
            }

            var result = _reportService.GenerateReport(json, template);

            Snapshot.Match(result);
        }
    }
}