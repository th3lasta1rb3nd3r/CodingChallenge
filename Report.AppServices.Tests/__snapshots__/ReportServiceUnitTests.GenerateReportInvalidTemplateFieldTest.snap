[
  {
    "FieldName": "lastname1",
    "Message": "Invalid template field"
  }
]
